drivers: video: gc2145: Include set_resolution() in the format caps check

josuah opened the pull request:
The current GalxyCore GC2145 does not guard against a wrong pixelformat passed.

This usually never happens and might be caught by the `gc2145_set_output_format()` anyway, but this permits to only set `drv_data->fmt` if the format is valid.

A very minor change!

josuah committed:
drivers: video: gc2145: check format compatibility before applying

While applying the format, the pixel format and drv_data->fmt were set
immediately, and the resolution was set only if it had a matching
entry on the "caps".

This commit makes sure the requested format matches the caps before
applying the format as well as drv_data->fmt. This does not guards
against partial failure (i.e. only pixelformat set and not
resolution).

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: gc2145: return 0 at the end of functions

After the error code is checked to be zero, it is possible to return 0
explicitly to help with readability. Also, when available, forward the
return code from the failing function instead of a locally chosen error
code like -EIO.

Signed-off-by: Josuah Demangeon <me@josuah.net>

uLipe reviewed (dismissed):
LGTM, it is perfectly reasonable grouping the I/O operations only after making sure the the format is supported.

pillo79 reviewed (dismissed):
ACK, though if you are up for another round... in Zephyr usually failures are handled in the `if`s, and the function returns at the end if everything is OK. You could refactor `set_fmt` so that if it does not find a match it returns an error.

uLipe commented:
@pillo79 , 

> ACK, though if you are up for another round... in Zephyr usually failures are handled in the `if`s, and the function returns at the end if everything is OK. You could refactor `set_fmt` so that if it does not find a match it returns an error.

@pillo79 , most of these drivers were written and their logic got oversight after merging, is there a proposal of refactoring them to make them consistent of the Zephyr standard, I have backlog here that I plan to start once I finish my work with the Arduino Nicla, and refactor one driver by one.

pillo79 commented:
> > ACK, though if you are up for another round... in Zephyr usually failures are handled in the `if`s, and the function returns at the end if everything is OK. You could refactor `set_fmt` so that if it does not find a match it returns an error.
> 
> @pillo79 , most of these drivers were written and their logic got oversight after merging, is there a proposal of refactoring them to make them consistent of the Zephyr standard, I have backlog here that I plan to start once I finish my work with the Arduino Nicla, and refactor one driver by one.

Sure, that's great and that's why I approved in the first place :slightly_smiling_face:  
I just noted that this PR is actually making the above point slightly worse, in the sense that it moves _more_ code inside the `if`. My POV was simply "if you're gonna touch it, might as well fix it once". Not a deal breaker at all!

josuah commented:
> is there a proposal of refactoring them to make them consistent of the Zephyr standard

Kind of, maybe the first step is to agree on a "ideal format".
This is the goal of this:
- https://github.com/zephyrproject-rtos/zephyr/issues/73867

I was tempted to start with the GC2145 as it got a review round directly from a TSC member here:
- https://github.com/zephyrproject-rtos/zephyr/pull/77770

The first sensors were contributed a long time ago, and it is good practice to pick existing drivers to write new ones.
This leads to a lot of refactoring to do for each and every new sensor to make it look closer to recent Zephyr guidelines.

josuah committed:
drivers: video: gc2145: set_fmt: branch on failure rather than success

This aims to make the code more linear by having the for loop
validates the input format rather than search for a match.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Force push: the new driver passes the `enum resolutions` to `set_resolution` instead of the width/height, so the build did fail. Remove the `width` and `height` variables set but not used after the refactoring.

pillo79 commented on drivers/video/gc2145.c:
```diff
@@ -1032,14 +1032,24 @@ static int gc2145_set_fmt(const struct device *dev, enum video_endpoint_id ep,
 	}
 
+	/* Check if camera is capable of handling given format */
+	for (int i = 0;; i++) {
+		if (i == ARRAY_SIZE(fmts)) {
+			LOG_ERR("Image format not supported\n");
+			return -ENOTSUP;
+		}
+		if (fmts[i].width_min == fmt->width && fmts[i].height_min == fmt->height &&
+		    fmts[i].pixelformat == fmt->pixelformat) {
+			res = (enum resolutions)i;
+			break;
+		}
+	}
```
Nit: I think a more common way to write this would be:
```C
	enum resolutions res = RESOLUTIONS_MAX;
	
	for (int i = 0; i < ARRAY_SIZE(fmts); i++) {
		if (fmts[i].width_min == fmt->width && fmts[i].height_min == fmt->height &&
		    fmts[i].pixelformat == fmt->pixelformat) {
			res = (enum resolutions)i;
			break;
		}
	}
	if (res == RESOLUTIONS_MAX) {
		LOG_ERR("Image format not supported\n");
		return -ENOTSUP;
	}
```

josuah replied:
Then there is the more obvious `for (i = 0; i < N; i++)` idiom standing out.
I appreciate the teachings. :)

pillo79 reviewed (dismissed):
ACK - though the new `for` is quirky :slightly_smiling_face: 
Up to you if you wish to amend with something like my comment.
